src/libsyntax/parse/parser/pat.rs

```diff
@@ -97,6 +100,30 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
+    /// Parses a pattern, that may be a or-pattern (e.g. `Some(Foo | Bar)`).
+    fn parse_pat_with_or(&mut self, expected: Option<&'static str>) -> PResult<'a, P<Pat>> {
+        let first_pat = self.parse_pat(expected)?;
+        if self.check(&token::BinOp(token::Or)) {
```

Comment:
```suggestion
        if !self.check(&token::BinOp(token::Or)) {
            return Ok(first_pat)
        }

```
Returning early seems overall cleaner here.